File: lld/week-02-03-patterns/decorator/python/decorator.py
"""Decorator pattern — a base Coffee wrapped by stackable add-on decorators.
See ../README.md for the design writeup."""
from __future__ import annotations

from typing import Protocol


class Coffee(Protocol):
    def cost(self) -> float: ...
    def description(self) -> str: ...


class Espresso:
    def cost(self) -> float:
        return 2.0

    def description(self) -> str:
        return "Espresso"


class _CoffeeDecorator:
    def __init__(self, wrapped: Coffee):
        self._wrapped = wrapped


class MilkDecorator(_CoffeeDecorator):
    def cost(self) -> float:
        return self._wrapped.cost() + 0.5

    def description(self) -> str:
        return self._wrapped.description() + " + Milk"


class SugarDecorator(_CoffeeDecorator):
    def cost(self) -> float:
        return self._wrapped.cost() + 0.25

    def description(self) -> str:
        return self._wrapped.description() + " + Sugar"


class WhipDecorator(_CoffeeDecorator):
    def cost(self) -> float:
        return self._wrapped.cost() + 0.75

    def description(self) -> str:
        return self._wrapped.description() + " + Whip"


def _demo() -> None:
    coffee: Coffee = Espresso()
    coffee = MilkDecorator(coffee)
    coffee = SugarDecorator(coffee)
    coffee = WhipDecorator(coffee)
    print(f"{coffee.description()} costs {coffee.cost()}")


if __name__ == "__main__":
    _demo()

File: lld/week-02-03-patterns/decorator/go/decorator.go
// Package decorator demonstrates the Decorator structural pattern:
// a base Coffee is wrapped by add-on decorators (milk, sugar, whip) that
// each layer in additional cost and description without subclass explosion.
package decorator

// Coffee is the component interface shared by the base drink and all decorators.
type Coffee interface {
	Cost() float64
	Description() string
}

type Espresso struct{}

func (Espresso) Cost() float64        { return 2.0 }
func (Espresso) Description() string  { return "Espresso" }

// MilkDecorator wraps a Coffee, adding its own cost/description on top.
type MilkDecorator struct {
	Wrapped Coffee
}

func (d MilkDecorator) Cost() float64 { return d.Wrapped.Cost() + 0.5 }
func (d MilkDecorator) Description() string {
	return d.Wrapped.Description() + " + Milk"
}

type SugarDecorator struct {
	Wrapped Coffee
}

func (d SugarDecorator) Cost() float64 { return d.Wrapped.Cost() + 0.25 }
func (d SugarDecorator) Description() string {
	return d.Wrapped.Description() + " + Sugar"
}

type WhipDecorator struct {
	Wrapped Coffee
}

func (d WhipDecorator) Cost() float64 { return d.Wrapped.Cost() + 0.75 }
func (d WhipDecorator) Description() string {
	return d.Wrapped.Description() + " + Whip"
}
